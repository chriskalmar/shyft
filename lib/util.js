
import DataType from './datatype/DataType';
import Entity from './entity/Entity';


export const passOrThrow = (condition, messageFn) => {
  // providing the error message as a callback return massively improves code coverage reports.
  // the message function is only evaluated if the condition fails, which will show up correctly
  // in the coverage reports.
  if (typeof messageFn !== 'function') {
    throw new Error('passOrThrow() expects messageFn to be a function')
  }
  if (!condition) {
    throw new Error(messageFn())
  }
}


export const isDataType = (obj) => {
  return (obj instanceof DataType)
}


export const isEntity = (obj) => {
  return (obj instanceof Entity)
}


export const resolveFunctionMap = (functionOrMap) => {
  return typeof functionOrMap === 'function'
    ? functionOrMap()
    : functionOrMap
}


export const isMap = (map, nonEmpty) => {
  return map !== null &&
    typeof map === 'object' &&
    Array.isArray(map) === false &&
    (!nonEmpty || (nonEmpty && Object.keys(map).length > 0 ))
}


export const isFunction = (fn) => {
  return typeof fn === 'function'
}



export const isArray = (set, nonEmpty) => {
  return set !== null &&
    typeof map === 'array' &&
    Array.isArray(set) === true &&
    (!nonEmpty || (nonEmpty && set.length > 0 ))
}

