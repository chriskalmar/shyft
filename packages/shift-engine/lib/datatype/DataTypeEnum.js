
import {
  passOrThrow,
  isArray,
} from '../util';

import DataType from './DataType';


class DataTypeEnum extends DataType {

  constructor (setup = {}) {

    const {
      name,
      description,
      values,
    } = setup

    passOrThrow(
      isArray(values, true),
      () => `'Missing enum values for data type '${name}'`
    )

    super({
      ...setup,
      description: description || `Enumeration set: ${values.join(', ')}`,
      mock() {
        const randomPos = Math.floor( Math.random() * values.length )
        return values[ randomPos ]
      }
    })

    this.values = values
  }


  toString() {
    return this.name
  }

}


export default DataTypeEnum


export const isDataTypeEnum = (obj) => {
  return (obj instanceof DataTypeEnum)
}
