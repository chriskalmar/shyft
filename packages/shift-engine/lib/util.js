import casual from 'casual';
import _ from 'lodash';



export const deleteUndefinedProps = (obj) =>
  Object.keys(obj).forEach(key => obj[key] === undefined && delete obj[key])



export const passOrThrow = (condition, messageFn) => {
  // providing the error message as a callback return massively improves code coverage reports.
  // the message function is only evaluated if the condition fails, which will show up correctly
  // in the coverage reports.
  if (typeof messageFn !== 'function') {
    throw new Error('passOrThrow() expects messageFn to be a function')
  }
  if (!condition) {
    throw new Error(messageFn())
  }
}


export const resolveFunctionMap = (functionOrMap) => {
  return typeof functionOrMap === 'function'
    ? functionOrMap()
    : { ...functionOrMap }
}


export const isMap = (map, nonEmpty) => {
  return map !== null &&
    typeof map === 'object' &&
    Array.isArray(map) === false &&
    (!nonEmpty || (nonEmpty && Object.keys(map).length > 0 ))
}


export const isFunction = (fn) => {
  return typeof fn === 'function'
}



export const isArray = (set, nonEmpty) => {
  return set !== null &&
    Array.isArray(set) === true &&
    (!nonEmpty || (nonEmpty && set.length > 0 ))
}


export const isString = (str) => (typeof str === 'string')


export const mergeMaps = (first, second) => {
  if (!isMap(first) || !isMap(second)) {
    throw new Error('mergeMaps() expects 2 maps for a merge to work')
  }

  return { ...first, ...second }
}



export const mapOverProperties = (object, iteratee) => {

  passOrThrow(
    isMap(object),
    () => 'Provided object is not a map'
  )

  passOrThrow(
    isFunction(iteratee),
    () => 'Provided iteratee is not a function'
  )


  const keys = Object.keys(object);

  keys.forEach((key) => {
    iteratee(object[ key ], key)
  })
}



export const sortDataByKeys = (keys, data, keyProperty='id') => {
  const map = {}
  const result = []
  const order = {}

  if (!keys || ( isArray(keys) && keys.length === 0)) {
    return []
  }

  if (!data || ( isArray(data) && data.length === 0)) {
    return keys.map(() => null)
  }


  for (let d = 0; d < data.length; d++) {
    const row = data[ d ]
    const id = row[ keyProperty ]
    const found = map[id]

    if (!found) {
      map[id] = row
    }
    else if (found instanceof Array) {
      found.push(row)
    }
    else {
      map[id] = [ found, row ];
    }
  }


  for (let k = 0; k < keys.length; k++) {
    const key = keys[ k ]
    const found = map[ key ];

    if (typeof found === 'undefined') {
      result.push(null)
    }
    else if (found instanceof Array) {
      let idx = order[ key ];

      if (typeof idx === 'undefined') {
        idx = 0
      }
      else {
        idx++
      }

      order[ key ] = idx;

      result.push(found[ idx ]);
    }
    else {
      result.push(found);
    }
  }

  return result;
};



export const processCursor = (entity, cursor, orderBy, reverse) => {

  const $LT = reverse ? '$gt' : '$lt'
  const $GT = reverse ? '$lt' : '$gt'
  const $LTE = reverse ? '$gte' : '$lte'
  const $GTE = reverse ? '$lte' : '$gte'

  const where = {}

  if (cursor) {
    passOrThrow(
      isArray(cursor[ entity.name ]),
      () => 'Incompatible cursor for this entity'
    )

    passOrThrow(
      isArray(orderBy),
      () => 'orderBy needs to be an array of order definitions'
    )

    const orderList = []

    const orderMap = {}
    orderBy.map(({attribute, direction}) => {

      passOrThrow(
        attribute && direction,
        () => 'orderBy needs to be an array of attributes and respective sort order'
      )

      orderMap[ attribute ] = direction
    })

    const primaryAttribute = entity.getPrimaryAttribute()
    const attributes = entity.getAttributes()

    let foundUniqueAttribute = false

    cursor[ entity.name ].map(filter => {
      if (filter.length !== 2) {
        throw new Error('Cursor malformed')
      }

      const attributeName = filter[0]

      passOrThrow(
        attributes[ attributeName ],
        () => `Unknown attribute '${attributeName}' used in cursor`
      )

      if (attributeName !== primaryAttribute.name) {
        passOrThrow(
          orderMap[ attributeName ],
          () => `Cursor works only on sorted attributes (check: '${attributeName}')`
        )
      }

      const attribute = attributes[ attributeName ]

      // limit where clause to the first attribute which is defined as unique
      if (!foundUniqueAttribute) {
        orderList.push(attributeName)
      }

      if (attribute.isUnique) {
        foundUniqueAttribute = true
      }

    })

    passOrThrow(
      foundUniqueAttribute,
      () => 'Cursor needs to have at least one attribute defined as unique'
    )


    // if more than 2 attributes are used for the cursor take only the first and the last (primary key)
    if (orderList.length > 2) {
      orderList.splice(1, (orderList.length - 2) )
    }


    // simple filter for single attributes
    if (orderList.length === 1) {

      const attributeName = orderList[0]
      const value = cursor[ entity.name ][0][1]

      if (orderMap[ attributeName ] === 'DESC') {
        where[ attributeName ] = {
          [ $LT ]: value
        }
      }
      else {
        where[ attributeName ] = {
          [ $GT ]: value
        }
      }

    }
    else {
      where.$not = {}

      cursor[ entity.name ].map(filter => {
        const attributeName = filter[0]
        const value = filter[1]

        // ignore attributes that obsolete due to prior unique attribute
        if (orderList.indexOf(attributeName) === -1) {
          return
        }

        const attribute = attributes[ attributeName ]

        if (attribute.isUnique) {
          if (orderMap[ attributeName ] === 'DESC') {
            where.$not[ attributeName ] = {
              [ $GTE ]: value
            }
          }
          else {
            where.$not[ attributeName ] = {
              [ $LTE ]: value
            }
          }
        }
        else {
          where.$not[ attributeName ] = value

          if (orderMap[ attributeName ] === 'DESC') {
            where[ attributeName ] = {
              [ $LTE ]: value
            }
          }
          else {
            where[ attributeName ] = {
              [ $GTE ]: value
            }
          }
        }

      })
    }
  }

  return where
}


export const processCursors = (entity, args) => {

  const {
    after,
    before,
  } = args;

  const where = {
    $and: [
      processCursor(entity, after, args.orderBy),
      processCursor(entity, before, args.orderBy, true),
    ]
  }

  return where
}



export const reverseString = (str) => {
  return str
    .split('')
    .reverse()
    .join('');
}



export const randomJson = () => {
  const ret = {}
  const count = casual.integer(1, 5)

  _.times(count, () => {
    const key = _.camelCase(casual.words(2))
    const value = Math.random() > 0.5
      ? casual.title
      : casual.integer()

    ret[key] = value
  })

  return ret
}
