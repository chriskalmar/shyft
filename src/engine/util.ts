import * as casual from 'casual';
import * as _ from 'lodash';
import { viewAttributePropertiesWhitelist } from './constants';

type StringFunction = () => string;

export const deleteUndefinedProps = (obj) =>
  Object.keys(obj).forEach((key) => obj[key] === undefined && delete obj[key]);

export const passOrThrow = (
  condition: any,
  messageFn: string | StringFunction,
): void => {
  // providing the error message as a callback return massively improves code coverage reports.
  // the message function is only evaluated if the condition fails, which will show up correctly
  // in the coverage reports.
  if (typeof messageFn !== 'function') {
    throw new Error('passOrThrow() expects messageFn to be a function');
  }
  if (!condition) {
    throw new Error(messageFn());
  }
};

export const resolveFunctionMap = (functionOrMap) => {
  return typeof functionOrMap === 'function'
    ? functionOrMap()
    : { ...functionOrMap };
};

export const isMap = (map: object, nonEmpty?: boolean): boolean => {
  return (
    map !== null &&
    typeof map === 'object' &&
    Array.isArray(map) === false &&
    (!nonEmpty || (nonEmpty && Object.keys(map).length > 0))
  );
};

export const isFunction = (fn: any): boolean => {
  return typeof fn === 'function';
};

export const isArray = (set: any[], nonEmpty?: boolean): boolean => {
  return (
    set !== null &&
    Array.isArray(set) === true &&
    (!nonEmpty || (nonEmpty && set.length > 0))
  );
};

export const isString = (str: any): boolean => typeof str === 'string';

export const mergeMaps = (first, second) => {
  if (!isMap(first) || !isMap(second)) {
    throw new Error('mergeMaps() expects 2 maps for a merge to work');
  }

  return { ...first, ...second };
};

export const mapOverProperties = (
  object: object,
  iteratee: (val: any, key: any) => any,
) => {
  passOrThrow(isMap(object), () => 'Provided object is not a map');

  passOrThrow(
    isFunction(iteratee),
    () => 'Provided iteratee is not a function',
  );

  const keys = Object.keys(object);

  keys.forEach((key) => {
    iteratee(object[key], key);
  });
};

export const sortDataByKeys = (
  keys: string[],
  data: object[],
  keyProperty = 'id',
) => {
  const map = {};
  const result = [];
  const order = {};

  if (!keys || (isArray(keys) && keys.length === 0)) {
    return [];
  }

  if (!data || (isArray(data) && data.length === 0)) {
    return keys.map(() => null);
  }

  for (let d = 0; d < data.length; d++) {
    const row = data[d];
    const id = row[keyProperty];
    const found = map[id];

    if (!found) {
      map[id] = row;
    } else if (found instanceof Array) {
      found.push(row);
    } else {
      map[id] = [found, row];
    }
  }

  for (let k = 0; k < keys.length; k++) {
    const key = keys[k];
    const found = map[key];

    if (typeof found === 'undefined') {
      result.push(null);
    } else if (found instanceof Array) {
      let idx = order[key];

      if (typeof idx === 'undefined') {
        idx = 0;
      } else {
        idx++;
      }

      order[key] = idx;

      result.push(found[idx]);
    } else {
      result.push(found);
    }
  }

  return result;
};

export const reverseString = (str) => {
  return str.split('').reverse().join('');
};

export const randomJson = () => {
  const ret = {};
  const count = casual.integer(1, 5);

  _.times(count, () => {
    const key = _.camelCase(casual.words(2));
    const value = Math.random() > 0.5 ? casual.title : casual.integer();

    ret[key] = value;
  });

  return ret;
};

export const asyncForEach = async (
  array: any[],
  callback: (val, idx, arr) => any,
) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array);
  }
};

export const isDefined = (val: any): boolean => typeof val !== 'undefined';

export const convertEntityToViewAttribute = (attribute) => {
  if (!_.isObject(attribute)) {
    throw new Error(
      'convertEntityToViewAttribute() expects an attribute as input',
    );
  }

  const newAttribute = {};

  viewAttributePropertiesWhitelist.map((prop) => {
    newAttribute[prop] = attribute[prop];
  });

  return newAttribute;
};

export const convertEntityToViewAttributesMap = (attributesMap) => {
  const newAttributesMap = {};

  if (!_.isObject(attributesMap)) {
    throw new Error(
      'convertEntityToViewAttributesMap() expects an attributes map as input',
    );
  }

  for (const [name, attribute] of Object.entries(attributesMap)) {
    newAttributesMap[name] = convertEntityToViewAttribute(attribute);
  }

  return newAttributesMap;
};
