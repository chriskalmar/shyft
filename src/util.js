import crypto from 'crypto';

// postgres limitation
const SYSTEM_NAME_MAX_LENGTH = 63;

export const generateIndexName = (entityName, attributes, suffix = '_idx') => {
  let ret = `${entityName}_${suffix}_`;
  const attributesChain = (attributes || []).join('_');
  const attributesSortedChain = (attributes || []).sort().join('_');

  if (ret.length + attributesChain.length > SYSTEM_NAME_MAX_LENGTH) {
    const hashLength = 10;
    const hash = crypto
      .createHash('sha256')
      .update(attributesSortedChain)
      .digest('hex')
      .substr(0, hashLength);

    const shortendAttributesChain = attributesChain.substr(
      0,
      SYSTEM_NAME_MAX_LENGTH - hashLength - ret.length - 1,
    );

    ret = `${entityName}_${shortendAttributesChain}_${hash}_${suffix}`;
  } else {
    ret = `${entityName}_${attributesChain}_${suffix}`;
  }

  return ret;
};

export const invertDirection = direction => {
  return direction === 'DESC' ? 'ASC' : 'DESC';
};

export const getLimit = (first, last) => {
  if (first >= 0) {
    return first;
  } else if (last >= 0) {
    return last;
  }

  return 10;
};

export const asyncForEach = async (array, callback) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array);
  }
};

export const quote = item => {
  if (!item) {
    throw new Error('quote() requires an input');
  }

  // special care for json pointers
  if (item.includes('->')) {
    const arr = item.split('->');
    arr[0] = `"${arr[0]}"`;
    return arr.join('->');
  }

  return `"${item}"`;
};
