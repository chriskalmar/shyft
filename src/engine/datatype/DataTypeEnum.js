import { passOrThrow, isArray } from '../util';

import { enumValueRegex, ENUM_VALUE_PATTERN } from '../constants';

import DataType from './DataType';

export class DataTypeEnum extends DataType {
  constructor(setup = {}) {
    const { name, description, values } = setup;

    passOrThrow(
      isArray(values, true),
      () => `'Missing enum values for data type '${name}'`,
    );

    values.map(value => {
      passOrThrow(
        enumValueRegex.test(value),
        () =>
          `Invalid enum value '${value}' for data type '${name}' (Regex: /${ENUM_VALUE_PATTERN}/)`,
      );
    });

    super({
      ...setup,
      description: description || `Enumeration set: ${values.join(', ')}`,
      mock() {
        const randomPos = Math.floor(Math.random() * values.length);
        return values[randomPos];
      },
    });

    this.values = values;
  }

  toString() {
    return this.name;
  }
}

export const isDataTypeEnum = obj => {
  return obj instanceof DataTypeEnum;
};
