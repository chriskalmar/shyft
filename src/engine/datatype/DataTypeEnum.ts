import { uniq } from 'lodash';
import { passOrThrow, isMap } from '../util';
import { enumValueRegex, ENUM_VALUE_PATTERN } from '../constants';
import { DataType } from './DataType';

export type ValueType = {
  [key: string]: number;
};

export type DataTypeEnumSetupType = {
  name: string;
  description?: string;
  values: ValueType;
};

export class DataTypeEnum extends DataType {
  values: ValueType;

  constructor(setup: DataTypeEnumSetupType = {} as DataTypeEnumSetupType) {
    const { name, description, values } = setup;

    passOrThrow(
      isMap(values, true),
      () => `'Missing enum values for data type '${name}'`,
    );

    const valueNames = Object.keys(values);
    const uniqueIds = [];

    valueNames.map((valueName) => {
      const valueId = values[valueName];
      uniqueIds.push(valueId);

      passOrThrow(
        enumValueRegex.test(valueName),
        () =>
          `Invalid value name '${valueName}' for data type '${name}' (Regex: /${ENUM_VALUE_PATTERN}/)`,
      );

      passOrThrow(
        valueId === Number(valueId) && valueId > 0,
        () =>
          `Value '${valueName}' for data type '${name}' has an invalid unique ID (needs to be a positive integer)`,
      );
    });

    passOrThrow(
      uniqueIds.length === uniq(uniqueIds).length,
      () =>
        `Each value defined for data type '${name}' needs to have a unique ID`,
    );

    super({
      name,
      description: description || `Enumeration set: ${valueNames.join(', ')}`,
      enforceIndex: true,
      mock() {
        const randomPos = Math.floor(Math.random() * uniqueIds.length);
        return uniqueIds[randomPos];
      },
    });

    this.values = values;
  }

  toString(): string {
    return this.name;
  }
}

export const isDataTypeEnum = (obj: any): boolean => {
  return obj instanceof DataTypeEnum;
};
