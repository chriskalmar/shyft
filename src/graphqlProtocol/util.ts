import * as _ from 'lodash';
import * as pluralize from 'pluralize';
import { RELAY_TYPE_PROMOTER_FIELD } from './protocolGraphqlConstants';
import { CustomError } from '../engine/CustomError';
import { isDefined } from '../engine/util';
import { getRegisteredEntityAttribute } from './registry';

export function generateTypeName(name) {
  return _.camelCase(name);
}

export function pascalCase(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

export function plural(str) {
  return pluralize.plural(str);
}

export function generateTypeNamePascalCase(name) {
  return pascalCase(generateTypeName(name));
}

export function generateTypeNamePlural(name) {
  return pluralize.plural(generateTypeName(name));
}

export function generateTypeNamePluralPascalCase(name) {
  return pascalCase(pluralize.plural(generateTypeName(name)));
}

export function generateTypeNameUpperCase(name) {
  return _.snakeCase(name).toUpperCase();
}

export const toBase64 = (value) =>
  new Buffer(value.toString()).toString('base64');
export const fromBase64 = (value) =>
  new Buffer(value.toString(), 'base64').toString();

export const serializeCursor = (cursor) => toBase64(JSON.stringify(cursor));
export const deserializeCursor = (cursor) => {
  try {
    return JSON.parse(fromBase64(cursor));
  } catch (e) {
    // TODO: check
    // should throw error, not return
    // https://github.com/graphql/graphql-js/issues/910
    return new Error('Invalid cursor provided');
  }
};

export const addRelayTypePromoterToInstance = (typeName, instance) => {
  if (!instance) {
    return instance;
  }

  instance[RELAY_TYPE_PROMOTER_FIELD] = typeName;
  return instance;
};

export const addRelayTypePromoterToList = (typeName, list) => {
  return list.map((instance) => {
    return addRelayTypePromoterToInstance(typeName, instance);
  });
};

export const addRelayTypePromoterToInstanceFn = (typeName) => {
  return (instance) => {
    return addRelayTypePromoterToInstance(typeName, instance);
  };
};

export const addRelayTypePromoterToListFn = (typeName) => {
  return (list) => {
    return addRelayTypePromoterToList(typeName, list);
  };
};

export const translateInstance = (
  entity,
  instance,
  { i18nLanguage, i18nDefaultLanguage, i18nLanguageDisableFallback },
) => {
  if (
    !instance ||
    !i18nLanguage ||
    i18nLanguage === i18nDefaultLanguage ||
    !entity.getI18nAttributeNames ||
    !entity.getI18nAttributeNames()
  ) {
    return instance;
  }

  const attributes = entity.getAttributes();
  const i18nAttributeNames = entity.getI18nAttributeNames();

  i18nAttributeNames.map((attributeName) => {
    const { required } = attributes[attributeName];
    const {
      fieldName: gqlFieldName,
      fieldNameI18n: gqlFieldNameI18n,
    } = getRegisteredEntityAttribute(
      entity.name,
      attributes[attributeName].name,
    );

    const translation = instance[gqlFieldNameI18n]
      ? instance[gqlFieldNameI18n][i18nLanguage]
      : undefined;

    if (i18nLanguageDisableFallback) {
      instance[gqlFieldName] =
        !isDefined(translation) && required
          ? new CustomError(
              `Translation for '${gqlFieldName}' not found`,
              'TranslationNotFoundError',
            )
          : translation;
    } else if (isDefined(translation)) {
      instance[gqlFieldName] = translation;
    }
  });

  return instance;
};

export const translateList = (entity, list, context) => {
  return list.map((instance) => {
    return translateInstance(entity, instance, context);
  });
};

export const translateInstanceFn = (entity, context) => {
  return (instance) => {
    return translateInstance(entity, instance, context);
  };
};

export const translateListFn = (entity, context) => {
  return (list) => {
    return translateList(entity, list, context);
  };
};

// emulate gql tag for prettier
export const gql = String.raw;
