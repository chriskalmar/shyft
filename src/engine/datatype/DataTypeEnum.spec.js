import DataTypeEnum, { isDataTypeEnum } from './DataTypeEnum';

import { passOrThrow } from '../util';

describe('DataTypeEnum', () => {
  it('should have a set of values', () => {
    let fn;

    fn = () => {
      // eslint-disable-next-line no-new
      new DataTypeEnum();
    };

    expect(fn).toThrowErrorMatchingSnapshot();

    fn = () => {
      // eslint-disable-next-line no-new
      new DataTypeEnum({
        values: [],
      });
    };

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should have a set of valid values', () => {
    let fn;

    fn = () => {
      // eslint-disable-next-line no-new
      new DataTypeEnum({
        values: [ '8' ],
      });
    };

    expect(fn).toThrowErrorMatchingSnapshot();

    fn = () => {
      // eslint-disable-next-line no-new
      new DataTypeEnum({
        values: [ ' abc ' ],
        name: 'test',
      });
    };

    expect(fn).toThrowErrorMatchingSnapshot();

    fn = () => {
      // eslint-disable-next-line no-new
      new DataTypeEnum({
        values: [ 'abc', 'def', 'hello-there' ],
        name: 'another',
      });
    };

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should have a name', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new DataTypeEnum({
        values: [ 'item' ],
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it("should return it's name", () => {
    const dataType = new DataTypeEnum({
      name: 'someDataTypeName',
      description: 'Just some description',
      values: [ 'item' ],
    });

    expect(dataType.name).toBe('someDataTypeName');
    expect(String(dataType)).toBe('someDataTypeName');
  });

  it('should have a fallback description', () => {
    const dataType = new DataTypeEnum({
      name: 'example',
      values: [ 'ACTION', 'COMEDY', 'DRAMA' ],
    });

    expect(dataType.description).toBe('Enumeration set: ACTION, COMEDY, DRAMA');
  });

  it('should have a generated mock function', () => {
    const values = [ 'ACTION', 'COMEDY', 'DRAMA' ];
    const dataType = new DataTypeEnum({
      name: 'example',
      values,
    });

    expect(values.includes(dataType.mock())).toBe(true);
  });

  describe('isDataTypeEnum', () => {
    it('should recognize objects of type DataTypeEnum', () => {
      const enum1 = new DataTypeEnum({
        name: 'enum1',
        values: [ 'ACTION', 'COMEDY', 'DRAMA' ],
      });

      const enum2 = new DataTypeEnum({
        name: 'enum2',
        values: [ 'APPLE', 'PEAR', 'CHERRY' ],
      });

      function fn() {
        passOrThrow(
          isDataTypeEnum(enum1) && isDataTypeEnum(enum2),
          () => 'This error will never happen',
        );
      }

      expect(fn).not.toThrow();
    });

    it('should recognize non-DataTypeEnum objects', () => {
      function fn() {
        passOrThrow(
          isDataTypeEnum({}) ||
            isDataTypeEnum(function test() {}) ||
            isDataTypeEnum(Error),
          () => 'Not a DataTypeEnum object',
        );
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });
});
