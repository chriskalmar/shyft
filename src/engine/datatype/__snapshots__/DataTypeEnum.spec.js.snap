// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTypeEnum isDataTypeEnum should recognize non-DataTypeEnum objects 1`] = `"Not a DataTypeEnum object"`;

exports[`DataTypeEnum should have a name 1`] = `"Missing data type name"`;

exports[`DataTypeEnum should have a set of valid values 1`] = `"Invalid enum value '8' for data type 'undefined' (Regex: /^[_a-zA-Z][_a-zA-Z0-9]*$/)"`;

exports[`DataTypeEnum should have a set of valid values 2`] = `"Invalid enum value ' abc ' for data type 'test' (Regex: /^[_a-zA-Z][_a-zA-Z0-9]*$/)"`;

exports[`DataTypeEnum should have a set of valid values 3`] = `"Invalid enum value 'hello-there' for data type 'another' (Regex: /^[_a-zA-Z][_a-zA-Z0-9]*$/)"`;

exports[`DataTypeEnum should have a set of values 1`] = `"'Missing enum values for data type 'undefined'"`;

exports[`DataTypeEnum should have a set of values 2`] = `"'Missing enum values for data type 'undefined'"`;
